bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/evaluation/EvaluationInteractor.kt

```diff
@@ -0,0 +1,74 @@
+package io.bucketeer.sdk.android.internal.evaluation
+
+import android.annotation.SuppressLint
+import android.content.SharedPreferences
+import androidx.annotation.VisibleForTesting
+import io.bucketeer.sdk.android.internal.Constants
+import io.bucketeer.sdk.android.internal.evaluation.db.EvaluationDao
+import io.bucketeer.sdk.android.internal.logd
+import io.bucketeer.sdk.android.internal.loge
+import io.bucketeer.sdk.android.internal.model.Evaluation
+import io.bucketeer.sdk.android.internal.model.User
+import io.bucketeer.sdk.android.internal.remote.ApiClient
+import io.bucketeer.sdk.android.internal.remote.GetEvaluationsResult
+import java.util.concurrent.Executor
+
+internal class EvaluationInteractor(
+  private val apiClient: ApiClient,
+  private val evaluationDao: EvaluationDao,
+  private val sharedPrefs: SharedPreferences,
+  private val executor: Executor,
+) {
+  // key: userId
+  @VisibleForTesting
+  internal val latestEvaluations = mutableMapOf<String, List<Evaluation>>()
+
+  @VisibleForTesting
+  internal var currentEvaluationsId: String
+    get() = sharedPrefs.getString(Constants.PREFERENCE_KEY_USER_EVALUATION_ID, "") ?: ""
+    @SuppressLint("ApplySharedPref")
+    set(value) {
+      sharedPrefs.edit()
+        .putString(Constants.PREFERENCE_KEY_USER_EVALUATION_ID, value)
+        .commit()
+    }
+
+  @Suppress("MoveVariableDeclarationIntoWhen")
+  fun fetch(user: User, timeoutMillis: Long?): GetEvaluationsResult {
+    val currentEvaluationsId = this.currentEvaluationsId
+
+    val result = apiClient.getEvaluations(user, currentEvaluationsId, timeoutMillis)
+
+    when (result) {
+      is GetEvaluationsResult.Success -> {
+        val response = result.value.data
+        val newEvaluationsId = response.user_evaluations_id
+        if (currentEvaluationsId == newEvaluationsId) {
+          logd { "Nothing to sync" }
+          return result
+        }
+
+        val newEvaluations = response.evaluations.evaluations
+
+        val success = evaluationDao.deleteAllAndInsert(user.id, newEvaluations)
+        if (!success) {
+          loge { "Failed to update latest evaluations" }
+          return result
+        }
+
+        this.currentEvaluationsId = newEvaluationsId
+
+        latestEvaluations[user.id] = newEvaluations
+      }
+      is GetEvaluationsResult.Failure -> {
+        logd(result.error) { "ApiError: ${result.error.message}" }
+      }
+    }
+    return result
+  }
+
+  fun getLatest(userId: String, featureId: String): Evaluation? {
+    val evaluations = latestEvaluations[userId] ?: emptyList()
```

Comment:
Just confirming.
For example, when the `BKTClient#boolVariation` is called, it will call `EvaluationInteractor#getLatest`, right?
If it returns null, inside of `BKTClient#boolVariation`, will it try to return the data from the DB?

Reply:
Thanks, I almost forgot about that.
That's one of the most complex parts of the lib.

Since `BKTClient#boolVariation` and others are synchronous methods and it can be called from the main thread, we can't simply return data from the DB. It can block the thread.

So we need to ensure that

- data is loaded from the DB on SDK start-up
- cache in EventInteractor is always up-to-date(make sure to update the cache when we update the DB)

At first, I think I could do some thing like below in `BKTClient#initialize`

```kotlin
val instance: BKTClient

fun initialize(
  context: Context,
  config: BKTConfig,
  user: BKTUser,
  timeoutMillis: Long = 5000
): Future<BKTClient> {
  if (instance != null) {
    logw { "BKTClient is already initialized. not sure if initial fetch has been finished" }
    return Futures.success(instance!!)
  }

  val client = BKTClientImpl(context, config, user)

  instance = client

  // fun initializeCache(): Future<Unit>
  // asynchronously load data from the DB, then commence initial fetch
  return client.initializeCache()
    .then { client.fetchEvaluations() }
  
}
```

But this isn't possible until Java8

Reply:
May be possbile

```kotlin
val instance: BKTClient

fun initialize(
  context: Context,
  config: BKTConfig,
  user: BKTUser,
  timeoutMillis: Long = 5000
): Future<BKTClient> {
  if (instance != null) {
    logw { "BKTClient is already initialized. not sure if initial fetch has been finished" }
    return Futures.success(instance!!)
  }

  val client = BKTClientImpl(context, config, user)

  instance = client

  return client.executorService.submit {
    client.initializeCache()
    // synchronous version of fetchEvaluations 
    client.fetchEvaluationsInternal()
  }
}
```